### PR TITLE
test: Code Coverage: Add Edit Mileage #5 Part 2

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 0b8d576</samp>

This pull request adds a new feature to create or edit mileage transactions in the `add-edit-mileage.page.ts` file. It also fixes a bug in the policy service that caused validation errors with locations data, and updates the interfaces and mock data for transactions and policy expenses to handle different formats of locations data. It also adds and imports a new test file for the add-edit-mileage page.

### <samp>🤖 Generated by Copilot at 0b8d576</samp>

> _We are the masters of mileage, we control the data flow_
> _We mock the transactions, we test the scenarios_
> _We change the interfaces, we handle the locations_
> _We fix the policy bugs, we create the `reportValue`_

### <samp>🤖 Generated by Copilot at 0b8d576</samp>

*  Add a new optional property `type` to the add-edit-mileage page class and constructor to store and receive the type of the transaction ([link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-04c216da6200af711aa0f53bcd7e253086762b5d123b0729fac3b1862ec5fc3eR2421), [link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-04c216da6200af711aa0f53bcd7e253086762b5d123b0729fac3b1862ec5fc3eR2438))
*  Add optional chaining operators to the `reportValue` object and its properties to handle null or undefined cases ([link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-04c216da6200af711aa0f53bcd7e253086762b5d123b0729fac3b1862ec5fc3eL2544-R2549))
*  Modify the interfaces for platform policy expense, public policy expense and transaction to allow the `locations` property to be either an array of destination objects, an array of strings, or an array of objects with a display property ([link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-d4b66927d68d2bc65589170acb4d89b7b3c233cbc69f6a01bce7dfbb2e72b0ddL22-R22), [link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-a0bc59def1dacbb0e0b8682fb52ecc114af35777a19b35355bf95d27f18a4958L41-R41), [link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-db96e89a4671a6577872d9c95a3722abc4d79e6b3f8443ec4f74cedb81915dbaL42-R42))
*  Cast the transaction `locations` property to an array of strings in the policy service and pass it to the policy engine for validation ([link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-d422f4879ab43f145be15ff4d400cfdf80d1c25a79012dc1c4073947149d913fR24), [link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-d422f4879ab43f145be15ff4d400cfdf80d1c25a79012dc1c4073947149d913fL42-R43))
*  Add new mock data objects and arrays for default transaction field values, organization category list items and unflattened transactions with different scenarios ([link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-88d03c9481ee9f872719d0aa13033ce30c0e50a9d5de1403d54b362a3f5282cbR31-R36), [link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-3f08ed1afe1b627dd8d79ee0eb6e1b895d7e1505e3670f46c43b97c46e016c9eR70-R132), [link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-21ae928ce640ad62cfada5fee7bf0bbc490300ab04a25f8309a9878e21b2f1cfR2837-R2859))
*  Add a new test file `add-edit-mileage.page.scenarios.spec.ts` that contains test cases for the add-edit-mileage page with different scenarios and import and invoke it in the add-edit-mileage page setup spec file ([link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-df40b6e711e89d5e80e6f3867394ee96bbff1eeae11eef42db09ce0b3f1821fbR61), [link](https://github.com/fylein/fyle-mobile-app/pull/2309/files?diff=unified&w=0#diff-df40b6e711e89d5e80e6f3867394ee96bbff1eeae11eef42db09ce0b3f1821fbR441))

## Clickup
https://app.clickup.com/t/85ztjmtfc

## Code Coverage
`50.11% Statements 423/844`
`44.84% Branches 322/718`
`45.91% Functions 135/294`
`50.87% Lines 409/804`

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/core/services/policy.service.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ export class PolicyService {
   ) {}
 
   transformTo(transaction: PublicPolicyExpense | Partial<Transaction>): PlatformPolicyExpense {
+    const locs = transaction.locations as string[];
```
Comment (Dimple16) [inline src/app/core/services/policy.service.ts:24]: const `locations`?

Comment (Dimple16) [reply src/app/core/services/policy.service.ts:24]: or `txnLocations`?

File: src/app/fyle/add-edit-mileage/add-edit-mileage-3.spec.ts [hunk 2]
```diff
@@ -391,5 +409,117 @@ export function TestCases3(getTestBed) {
...
+    it('trackEditExpense(): should track edit expense event', () => {
+      spyOn(component, 'getTimeSpentOnPage').and.returnValue(300);
+      component.presetProjectId = expectedProjectsResponse[0].project_id;
+      component.presetCostCenterId = costCentersData[0].id;
+      component.presetVehicleType = 'car';
+      component.presetLocation = 'Kolkata';
+      fixture.detectChanges();
+
+      component.trackEditExpense(unflattenedTxnWithTrackData);
+      expect(trackingService.editExpense).toHaveBeenCalledOnceWith({
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage-3.spec.ts:512]: I guess we have a model for this no? 
let us move it to that mock data file

**Review by Dimple16 — APPROVED** [review]
added comments

**Review by OmkarJ13 — APPROVED** [review]

**Pull request merged**